coil-fakeimageloader/src/main/java/coil/FakeImageLoader.kt

```diff
@@ -0,0 +1,232 @@
+        )
+
+        @Synchronized
+        fun enqueue(data: Any, engine: Engine) = apply {
+            results.getOrPut(data) { LinkedList() }.offer(engine)
+        }
+
+        @Synchronized
+        fun default(default: Engine) = apply {
+            this.default = default
+        }
+
+        override suspend fun execute(request: ImageRequest): ImageResult {
+            // NOTE: synchronized works here because we don't suspend inside the block.
+            val engine = synchronized(this) {
+                checkNotNull(results[request.data]?.poll() ?: default) {
+                    "no queued result for $request. results: $results"
+                }
+            }
+            return engine.execute(request)
+        }
+    }
+
+    private class JobDisposable(
+        override val job: Deferred<ImageResult>
+    ) : Disposable {
+
+        override val isDisposed: Boolean
+            get() = !job.isActive
+
+        override fun dispose() {
+            if (isDisposed) return
+            job.cancel()
+        }
+    }
+
+    class Builder {
+
+        private var coroutineContext = SupervisorJob() + Dispatchers.Unconfined
+        private var defaults = DefaultRequestOptions()
+        private var components = ComponentRegistry()
+        private var memoryCache: MemoryCache = FakeMemoryCache()
+        private var diskCache: DiskCache = FakeDiskCache()
+        private var engine: Engine = CompositeQueueEngine()
+
+        fun coroutineContext(context: CoroutineContext) = apply {
+            this.coroutineContext = context
+        }
+
+        fun defaults(defaults: DefaultRequestOptions) = apply {
+            this.defaults = defaults
+        }
+
+        fun components(components: ComponentRegistry) = apply {
+            this.components = components
+        }
+
+        fun memoryCache(memoryCache: MemoryCache) = apply {
+            this.memoryCache = memoryCache
+        }
+
+        fun diskCache(diskCache: DiskCache) = apply {
+            this.diskCache = diskCache
+        }
+
+        fun engine(engine: Engine) = apply {
+            this.engine = engine
+        }
+
+        fun build() = FakeImageLoader(
+            coroutineContext = coroutineContext,
+            defaults = defaults,
+            components = components,
+            memoryCache = memoryCache,
+            diskCache = diskCache,
+            engine = engine,
+        )
+    }
+}
+
+/**
+ * Create a new [FakeImageLoader] without configuration.
+ */
+@JvmName("create")
+fun FakeImageLoader(): FakeImageLoader {
+    return FakeImageLoader.Builder().build()
+}
+
+/** Convenience function to get [FakeImageLoader.memoryCache] as a [FakeMemoryCache]. */
+val FakeImageLoader.fakeMemoryCache: FakeMemoryCache get() = memoryCache as FakeMemoryCache
+
+/** Convenience function to get [FakeImageLoader.diskCache] as a [FakeDiskCache]. */
+val FakeImageLoader.fakeDiskCache: FakeDiskCache get() = diskCache as FakeDiskCache
+
+fun FakeImageLoader.enqueue(
```

Comment:
enqueue is an awkward name given the ImageLoader.enqueue also. one is a push, one is a pull.

Reply:
+1, pulled the name from `MockWebServer`, but there's definitely a better name for this case. Maybe `offer` or `push`?

Reply:
I suck at naming things. Those work, or enqueueResponse.